### ISSUE TITLE
Add ability to add metafields with other objects

### ISSUE THREAD

**Issue by MisinformedDNA** [nozzlegear/ShopifySharp#230]
Issue #135 removed the metafields property from all applicable objects. Due to that change, there is no way to bulk add metafields when connecting from an existing product store. Due to API rate limiting, being able to bulk add saves immense amounts of actual running time.

**Comment by nozzlegear** [comment]
Thanks for filing this. I've read your comment on #135 too and I understand the frustration. I wanted to remove Metafield fields from objects because they would only ever be used when creating or updating, which meant they'd always be useless empty lists when getting or listing an object. It could have certainly been solved by adding some documentation that said they're only for creating/updating, but I believe a better solution is to just add overloads to any create or update calls for objects that support metafields:

```cs
public async Task<SomeObject> Create(SomeObject obj, IEnumerable<MetaField> metaFields = null)
{
    if (metaFields != null && metaFields.Count > 0)
    {
        // Add metafields to request body here
    }
}
```

Sadly I never got that in before 4.0 -- I don't use metafields myself and haven't put much personal effort into supporting them because of that. Thankfully adding overloads isn't a breaking change and we can get this in at any time. If you'd like to take a crack at it I'd be happy to merge a pull request, otherwise I'll get this in as soon as I have the time.

**Comment by dkershner6** [comment]
@nozzlegear - I am happy to write this, as we need it, and am going to do it anyway so happy to share.  BUT, some points:
- On a product create, for the product metafields it would be easy to use a IEnumerable as above, but would likely be very difficult to use the same for the variants and images of the product.  You would need some sort of identifier to match on (ID doesn't exist yet) and lay them out in KeyValuePair<string, IEnumerable<Metafield>>.
- Webhook bodies can contain metafields, and if you want to Deserialize or FromObject into a class, you would need a target class that contains metafields.
- If specified in the fields, certain responses (GET) could also contain metafields, causing a potential exception (although this person would likely know that).

I definitely want to make sure whatever path we go down is the one with the longest term benefit.

Proposal:
- Reconsider the distaste for an empty IEnumerable in most use cases.
OR
- We should create very similar classes, but with Metafields included <ProductWithMetafields>.  Would save the empty IE's, but cause additional updating each time we change a class.

Please let me know your thoughts,
DK

**Comment by nozzlegear** [comment]
Interesting, I wasn’t aware you could get metafields back when requesting the field, and I also didn’t know that webhooks could contain them. Again I don’t think I’ve ever had the chance/need to really use metafields yet, so I’m pretty out-of-touch when it comes to them. Since the lists would actually be used (and not empty) at some point, I’m 100% okay with adding the metafield properties back to any object that can support it.

--
Josh Harms

**Comment by dkershner6** [comment]
Sounds good, @nozzlegear .

I am a bit new to this library, so I may need some guidance, but I will get started on this and keep this issue in the loop.

Thank you,
DK

**Comment by nozzlegear** [comment]
Great! Just ping me if you have any questions, happy to help.

--
Josh Harms

**Comment by dkershner6** [comment]
@nozzlegear 

Pull request created: https://github.com/nozzlegear/ShopifySharp/pull/265

I will get to some other entities when I have time to properly test them.

**Comment by clement911** [comment]
Hi guys,

I agree that adding back the Metafields properties is the best way to handle this.

@dkershner6, I'm curious also about having in the metafields in webhook bodies and get requests.
I took a look at https://help.shopify.com/api/reference/events/webhook and I see that there is a `metafield_namespaces`.
 Is that how you are achieving this for webhooks?

How about GET request? What syntax do you use in the fields property to include metafields?

**Comment by dkershner6** [comment]
You are correct on the webhook usage.

The metafields in the GET requests is obviously undocumented, so ymmv on where and how it works, but they do appear to be headed in that direction.

From memory, it's something to the effect of metafields[namespace]. Similar to some of the older APIs like image metafields or collection sorting. @clement11

**Comment by clement911** [comment]
hmm ok. regarding the webhooks, do you know if there is a way to say 'send me all metafields'?

**Comment by dkershner6** [comment]
@clement911

They specifically don't allow "all" metafield calls due to their nature. Since any app can use them, they can become pretty numerous if you get a bunch of apps going.

If you really wanted to do something like this, you could periodically call the metafields for a single product, store all the metafield namespaces, and re-create the webhook with that array in the metafields namespaces.

**Comment by clement911** [comment]
Thanks @dkershner6 , I did think about that but I'm trying to avoid polling.

**Comment by nozzlegear** [comment]
Published in 4.15.2!
